Lib/test/test_functools.py

```diff
@@ -1158,6 +1161,264 @@ def __eq__(self, other):
+        # not all elements involved in a cycle
+        self._assert_cycle({1: {2}, 2: {3}, 3: {1}, 5: {4}, 4: {6}}, [1, 3, 2, 1])
+        # Multiple cycles
+        self._assert_cycle({1: {2}, 2: {1}, 3: {4}, 4: {5}, 6: {7}, 7: {6}},
+                           [1, 2, 1])
+        # Cycle in the middle of the graph
+        self._assert_cycle({1: {2}, 2: {3}, 3: {2, 4}, 4: {5}}, [3, 2])
+
+    def test_calls_before_preapare(self):
+        ts = functools.TopologicalSorter()
+
+        with self.assertRaisesRegex(ValueError, r"prepare\(\) must be called first"):
+            ts.get_ready()
+        with self.assertRaisesRegex(ValueError, r"prepare\(\) must be called first"):
+            ts.done(3)
+        with self.assertRaisesRegex(ValueError, r"prepare\(\) must be called first"):
+            ts.is_active()
+
+    def test_prepare_multiple_times(self):
+        ts = functools.TopologicalSorter()
+        ts.prepare()
+        with self.assertRaisesRegex(ValueError, r"cannot prepare\(\) more than once"):
+            ts.prepare()
+
+    def test_invalid_nodes_in_done(self):
+        ts = functools.TopologicalSorter()
+        ts.add(1, 2, 3, 4)
+        ts.add(2, 3, 4)
+        ts.prepare()
+        ts.get_ready()
+
+        with self.assertRaisesRegex(ValueError, "node 2 was not passed out"):
+            ts.done(2)
+        with self.assertRaisesRegex(ValueError, r"node 24 was not added using add\(\)"):
+            ts.done(24)
+
+    def test_done(self):
+        ts = functools.TopologicalSorter()
+        ts.add(1, 2, 3, 4)
+        ts.add(2, 3)
+        ts.prepare()
+
+        self.assertEqual(ts.get_ready(), (3, 4))
+        # If we don't mark anything as done, get_ready() returns nothing
+        self.assertEqual(ts.get_ready(), ())
+        ts.done(3)
+        # Now 2 becomes available as 3 is done
+        self.assertEqual(ts.get_ready(), (2,))
+        self.assertEqual(ts.get_ready(), ())
+        ts.done(4)
+        ts.done(2)
+        # Only 1 is missing
+        self.assertEqual(ts.get_ready(), (1,))
+        self.assertEqual(ts.get_ready(), ())
+        ts.done(1)
+        self.assertEqual(ts.get_ready(), ())
+        self.assertFalse(ts.is_active())
+
+    def test_is_active(self):
+        ts = functools.TopologicalSorter()
+        ts.add(1, 2)
+        ts.prepare()
+
+        self.assertTrue(ts.is_active())
+        self.assertEqual(ts.get_ready(), (2,))
+        self.assertTrue(ts.is_active())
+        ts.done(2)
+        self.assertTrue(ts.is_active())
+        self.assertEqual(ts.get_ready(), (1,))
+        self.assertTrue(ts.is_active())
+        ts.done(1)
+        self.assertFalse(ts.is_active())
+
+    def test_not_hashable_nodes(self):
+        ts = functools.TopologicalSorter()
+        self.assertRaises(TypeError, ts.add, dict(), 1)
+        self.assertRaises(TypeError, ts.add, 1, dict())
+        self.assertRaises(TypeError, ts.add, dict(), dict())
+
+    def test_order_of_insertion_does_not_matter(self):
```

Comment:
Hmm.  Order of insertion doesn't matter for this particular test case, because it forces a total order:  0 -> 1 -> 2 -> 3.  But, in general, order of insertion can matter!  For example, comment out the two `ts.add(2, 1)` lines in this test, and then there are two nodes without predecessors (0 and 2).  "The first" in the topsort is then (because dicts are ordered now) whichever of 0 and 2 was added first (which is 2 in the first block, but 0 in the second block).

Reply:
> order of insertion can matter!

Right, I think what I wanted to do here actually is something like `test_order_of_insertion_does_not_matter_between_groups`, so basically comparing the order in which the groups themselves come out of sorting (assuming no particular order in the groups).

Reply:
Modified the test in ed8c4ee. If you think you prefer no test at all, I can just remove it :)

Reply:
That would be cool.  But you need a different kind of generator then, like:

```python
def by_layer(ts):
    ts.prepare()
    while ts.is_active():
        chunk = ts.get_ready()
        yield set(chunk)
        ts.done(*chunk)
```

`tuple()` (or `list()`) applied to that should yield thingies that compare equal regardless of insertion order.

Reply:
I think this insight is important enough to include it in the documentation of `static_order` so I added a paragraph and an example in 0406781958

Reply:
> If you think you prefer no test at all,

Oh no!  Testing it is a good idea. 😃